### ISSUE TITLE
Plugin not working with Ableton Link

### ISSUE THREAD

**Issue by hellocatfood** [x42/mclk.lv2#1]
I'm trying to use this plugin in the Carla and Modep hosts to control a Korg Volca device.

When the host is using internal sync and the plugin is set to host sync I can start and stop the Volca easily using the plugin. But when I set the host to use Ableton Link sync mode the start/stop commands don’t work and tempo changes aren’t sent.

I can't work out if the problem is the plugin, Carla/Modep (and possibly by extension Mod devices) or the Ableton Link sync mode.

**Comment by hellocatfood** [comment]
Turns out I can get it working using the [poly] object https://forum.pdpatchrepo.info/topic/5311/polyphonic-voice-management-using-poly

**Comment by MarkusTratschitt** [comment]
You need the iPhone App SyncKontrol for Monotribe. With that app u turn ur Volca into a mobile device. Link is designed to sync Live with Mobiles. u need to connect microphone from the phone to the sync in of the volca then press play on the volca and then activate link in the app on the phone. Now start Ableton, press Link-Button and then u can control the bpm via the iPhone app. and ur devices stay in sync. this only works on iOS/MacOS. 
Thats the way I get my equipment in sync. Maybe that can help u to figure out the problem. I am new in programming, so I just can tell, what I do.

**Comment by x42** [comment]
The plugin either produces free-rolling MIDI-clock, or uses host-provided clock ([LV2 time](http://lv2plug.in/ns/ext/time) information).

The plugin itself is ignorant about Abelton Link, or JACK transport or LTC or any other means that the host might use for its internal clock.

@falktx  It seems that Carla does not (correctly) translate Abeton Link to LV2 BBT. Can you comment on that?

**Comment by falkTX** [comment]
I used it this just a few days ago, so Carla works afaik.

**Comment by x42** [comment]
@hellocatfood : which version of Carla are you using?

**Comment by hellocatfood** [comment]
I just updated earlier to carla to 2.2.0 and carla-git to 2.3.0-RC1 after seeing the announcement from falkTX's blog.

Sadly it's still not working in either 2.2.0 or 2.3.0. I recorded a video today of the problem

https://user-images.githubusercontent.com/151947/143205065-d1f69b7a-eed4-4a8a-89b3-aafecfa5b0ea.mp4

Sorry for all of the background noise, I just wanted to capture the sound of the synth so you could tell when it's working.

Hopefully the video shows that the midi clock plugin is only sending play/stop/transport information when not using ableton link.

**Comment by falkTX** [comment]
yeah. so a little extra tests would be useful.

1. if you put a midi monitor (like sherlock) after the mclk-out, does it show any data when link is enabled?
2. load step-sequencer and enable host-sync for it. does it play with link off and not play with link on as well?

**Comment by hellocatfood** [comment]
1.
Putting the Sherlock midi inspector after mclk-out shows the following when I press play:
`+000 F2 03 03    SongPosition     387    3`
And nothing else. When I press stop it shows:
```
+0000 FC            Stop
+0000 F2 F26 04 Song Position    550
+0000 F2 F26 04 Song Position    550
```
2.
I loaded the x42 step sequencer and it plays when I have it set to host-sync and have link both on and off. I didn't have it connected to mclk

**Comment by gtrainavicius** [comment]
This issue was raised in our [community](https://community.blokas.io/t/ableton-link-to-midi-broken/2970/4), so I looked into this, and I think I spotted the cause of the problem. The `self->mclk_last_tick` that's based on the current bar and beat assumes that it is going to be close to the frame number of the audio sample, however, when Ableton Link is used, the bar and beat can be quite large when the transport is started, and the sample number would be quite low, so `if (next_tick_offset >= n_samples) break;` prevents the plugin from generating any clock messages. While using internal syncing, these values are close to each other, so the clock messages are generated.

As an attempt to fix it, I've modified this line:

https://github.com/x42/mclk.lv2/blob/master/src/mclk.c#L489

To look like this:

```
double tmp;
self->mclk_last_tick = sample_position - modf(bb, &tmp) * samples_per_beat;
```

I assumed 'self->mclk_last_tick' is supposed to be the sample frame number that aligned with the last beat, but correct me if that's a wrong assumption. :)

Anyway, with this workaround, the clock messages are getting generated at the correct rate when in Ableton Link sync mode, and the BPM is matched. I didn't test however whether the beats themselves occur at the correct moment everywhere.
